pkg/blockbuilder/blockbuilder.go

```diff
@@ -242,142 +258,147 @@ func cycleEndAtStartup(interval, buffer time.Duration) time.Time {
 	return cycleEnd
 }
 
-type partitionLag struct {
-	Partition int32
-	Lag       int64
-
-	CommitRecTs  int64
-	SeenTillTs   int64
-	LastBlockEnd int64
-}
-
 // NextConsumeCycle manages consumption of currently assigned partitions.
 // The cycleEnd argument indicates the timestamp (relative to Kafka records) up until which to consume from partitions
 // in this cycle. That is, Kafka records produced after the mark will be consumed in the next cycle.
 func (b *BlockBuilder) NextConsumeCycle(ctx context.Context, cycleEnd time.Time) error {
-	b.assignmentMu.Lock()
-	assignment := b.assignment
-	b.assignmentMu.Unlock()
-	assignmentParts, ok := assignment[b.cfg.Kafka.Topic]
-	if !ok || len(assignmentParts) == 0 {
-		return fmt.Errorf("no partitions assigned in %+v, topic %s", assignment, b.cfg.Kafka.Topic)
-	}
-
 	defer func(t time.Time) {
 		b.metrics.consumeCycleDuration.Observe(time.Since(t).Seconds())
 	}(time.Now())
 
 	kadmClient := kadm.NewClient(b.kafkaClient)
 
-	lags, err := kadmClient.Lag(ctx, b.cfg.Kafka.ConsumerGroup)
-	if err != nil {
-		return fmt.Errorf("get consumer group lag: %w", err)
-	} else if err := lags.Error(); err != nil {
-		return fmt.Errorf("get consumer group lag: %w", err)
-	}
+	defer func() {
+		b.mu.Lock()
+		defer b.mu.Unlock()
 
-	// parts maps partition to this partition's current lag
-	parts := make(map[int32]partitionLag, len(assignmentParts))
-	groupLag := lags[b.cfg.Kafka.ConsumerGroup].Lag
-	for _, part := range assignmentParts {
-		pl, ok := groupLag.Lookup(b.cfg.Kafka.Topic, part)
+		if b.cancelActivePart != nil {
+			b.cancelActivePart(context.Canceled)
+			b.cancelActivePart = nil
+		}
+
+		b.parts.reset()
```

Comment:
I prefer this reset to be done just before the below loop. Feels safer and easier to read.